common/cache/lru.go

```diff
@@ -274,61 +266,66 @@ func (c *lru) putInternal(key interface{}, value interface{}, allowUpdate bool)
 	if c.maxSize == 0 {
 		return nil, nil
 	}
-	entrySize := getSize(value)
-	if entrySize > c.maxSize {
+	newEntrySize := getSize(value)
+	if newEntrySize > c.maxSize {
 		return nil, ErrCacheItemTooLarge
 	}
 
 	c.mut.Lock()
 	defer c.mut.Unlock()
 
-	c.currSize += entrySize
-	c.tryEvictUntilEnoughSpace()
-	// If there is still not enough space, remove the new entry size from the current size and return an error
-	if c.currSize > c.maxSize {
-		c.currSize -= entrySize
-		return nil, ErrCacheFull
+	if allowUpdate {
+		c.tryEvictUntilEnoughSpace(newEntrySize)
```

Comment:
for the update case, this seems wrong.. suppose you're replacing an item of size 100 with another of size 100. we don't need any extra space, but this will try to evict until we have 100 more. or replacing 100 with 50, we need 0. or if replacing 100 with 110, we only need 10 extra space but this will ask for 110.

for the update-not-allowed case too, we shouldn't try to evict anything if the element exists.

Reply:
discuss offline, this is not accuracy problem but very important for cache effective. I moved cache eviction under allowUpdate and related unittest. https://github.com/temporalio/temporal/pull/4698/files#diff-2acb946449ba572cc3fb38626184a7545190102ae4ec98a8ee39613a904e6698R449